axes/helpers.py

```diff
@@ -43,21 +45,22 @@ def get_cache_timeout() -> Optional[int]:
     for use with the Django cache backends.
     """
 
-    cool_off = get_cool_off()
+    cool_off = get_cool_off(username)
     if cool_off is None:
         return None
     return int(cool_off.total_seconds())
 
 
-def get_cool_off() -> Optional[timedelta]:
+def get_cool_off(username: Optional[str] = None) -> Optional[timedelta]:
```

Comment:
Would it be better to just supply the `request` object here so the user can use any field present in there? IP address or other metadata could be interesting as well. The request can also be used to carry attached attributes used in upstream / downstream processing in the middleware stack, if needed. `username` seems a bit limiting if we're implementing features here.

Adding the use of `request` object requires a few modifications in other places as well, but should be easy enough to test.

Reply:
That suggestion made a lot of sense when I read it, but now that I'm trying to make the changes, I remembered why I went for this initial implementation... 

The place where it's called, `user_login_failed` calls a function to resolve the username `get_client_username`, passing down some `credentials`. I can see that this function defaults to using `request.data` or `request.POST`.

I think I'm lacking a bit of context on the lifecycle where this is used, is that fair to assume that this will be called in a POST request where the username will be included in the body? It feels we might not be there in some cases like SSO/OAuth flow. Are these kind of flows out of the scope of django-axes?

Reply:
Hey, a good question on the authentication methods and flows, I initially missed this one. Sorry about that.

SSO flows such as SAML and OAuth are absolutely in scope of this plugin. Ideally django-axes should be login method agnostic, if possible.

If anyone else is checking this discussion it might be useful to see the Django authentication signals documentation at https://docs.djangoproject.com/en/dev/ref/contrib/auth/#module-django.contrib.auth.signals for reference.

You are right in that the `request` object might not contain all the relevant information needed to dig up e.g. username or other identifying login data if the login is coming in from e.g. a SAML request, but `request` is the only persistent information, really, that we get from all the login flows i.e. going up and down the middleware stack and in all the login signals as described at e.g. https://django-axes.readthedocs.io/en/latest/7_architecture.html

At the current time we always expect a `request` for the [`get_lockout_response`](https://github.com/jazzband/django-axes/pull/1222/files#diff-87f5ab498b667653f3c1f46b6a38cbe11ec0ffbad6d423a5a91c4055123ba218L445) callable and `credentials` as an optional. 

It actually might be nice to pass the `credentials` argument as well, as you pointed out.

If you're open to it I think it'd be OK to add that argument as well and bake a version 7.x from that. If you want to make a PR I'll be happy to make a new release for that.